File: crates/marmot-c/examples/smoke.zig
// End-to-end smoke test for the marmot-c ABI from Zig.
//
// Imports the real marmot.h via @cImport, so the generated header is the
// single source of truth for signatures and struct layout. Exercises the
// same surface as examples/smoke.c: client lifecycle, error + last-error
// paths, a DEEP walk of the recursive Markdown DTO tree, offline reads,
// the error taxonomy, best-effort identity, and the free-the-root-only
// ownership rule.
//
// The calls are wrapped in a small `Marmot` type whose methods return Zig
// error unions (`MarmotError`); a helper maps any non-OK status to an
// error and stashes the thread-local last-error text. Cleanup uses
// `defer client.deinit()`, the natural Zig resource idiom. No naked
// out-parameter arrays leak into the top-level flow.
//
// Build & run (see c-smoke.sh for the canonical invocation):
//   zig build-exe examples/smoke.zig -I include -lc -fllvm \
//       target/debug/libmarmot_c.a -lm -lpthread -ldl -lunwind \
//       -femit-bin=smoke_zig
//   ./smoke_zig <fresh-empty-home-dir>
//
// `-fllvm` selects the LLVM backend (a standard ELF the loader accepts);
// `-lunwind` supplies the `_Unwind_*` symbols the Rust staticlib needs
// (gcc/clang link these implicitly, Zig does not).

const std = @import("std");

const c = @cImport({
    @cInclude("marmot.h");
});

// --- error handling -------------------------------------------------------

const MarmotError = error{
    NullPointer,
    UnknownAccount,
    Failed,
};

// Thread-local last-error detail, copied out of the C string the runtime
// hands us (and freed) so callers can print it after the fact.
var last_error_buf: [1024]u8 = undefined;
var last_error_len: usize = 0;

fn stashLastError() void {
    const msg = c.marmot_last_error_message();
    if (msg == null) {
        last_error_len = 0;
        return;
    }
    const s = std.mem.sliceTo(msg, 0);
    const n = @min(s.len, last_error_buf.len);
    @memcpy(last_error_buf[0..n], s[0..n]);
    last_error_len = n;
    c.marmot_string_free(msg);
}

fn lastError() []const u8 {
    return if (last_error_len == 0) "(no detail)" else last_error_buf[0..last_error_len];
}

// Map a raw status into a Zig error, capturing the runtime's detail text
// on the failure path. OK is the only success value.
fn mapStatus(st: c.MarmotStatus) MarmotError!void {
    if (st == c.MARMOT_STATUS_OK) return;
    stashLastError();
    if (st == c.MARMOT_STATUS_NULL_POINTER) return MarmotError.NullPointer;
    if (st == c.MARMOT_STATUS_UNKNOWN_ACCOUNT) return MarmotError.UnknownAccount;
    return MarmotError.Failed;
}

// --- idiomatic wrapper over the raw client -------------------------------

const Marmot = struct {
    client: *c.MarmotClient,

    fn open(root: [*c]const u8, relays: []const [*c]const u8) MarmotError!Marmot {
        var client: ?*c.MarmotClient = null;
        try mapStatus(c.marmot_client_new(root, relays.ptr, relays.len, &client));
        return .{ .client = client.? };
    }

    fn deinit(self: Marmot) void {
        c.marmot_client_free(self.client);
    }

    fn isStopping(self: Marmot) MarmotError!bool {
        var stopping: bool = false;
        try mapStatus(c.marmot_client_is_stopping(self.client, &stopping));
        return stopping;
    }

    fn start(self: Marmot) MarmotError!void {
        try mapStatus(c.marmot_client_start(self.client));
    }

    fn shutdown(self: Marmot) MarmotError!void {
        try mapStatus(c.marmot_client_shutdown(self.client));
    }

    fn parseMarkdown(self: Marmot, text: [*c]const u8) MarmotError!*c.MarmotMarkdownDocument {
        var doc: ?*c.MarmotMarkdownDocument = null;
        try mapStatus(c.marmot_parse_markdown(self.client, text, &doc));
        return doc.?;
    }

    fn listAccounts(self: Marmot) MarmotError!*c.MarmotAccountSummaryList {
        var list: ?*c.MarmotAccountSummaryList = null;
        try mapStatus(c.marmot_list_accounts(self.client, &list));
        return list.?;
    }

    // Absent directory entries resolve to OK with a NULL string; the
    // caller owns and frees whatever comes back.
    fn npub(self: Marmot, account_id_hex: [*c]const u8) MarmotError![*c]u8 {
        var out: [*c]u8 = null;
        try mapStatus(c.marmot_npub(self.client, account_id_hex, &out));
        return out;
    }

    fn nip65Relays(self: Marmot, account_ref: [*c]const u8) MarmotError!*c.MarmotStringList {
        var out: ?*c.MarmotStringList = null;
        try mapStatus(c.marmot_account_nip65_relays(self.client, account_ref, &out));
        return out.?;
    }

    fn revealNsec(self: Marmot, account_ref: [*c]const u8) MarmotError![*c]u8 {
        var out: [*c]u8 = null;
        try mapStatus(c.marmot_reveal_nsec(self.client, account_ref, &out));
        return out;
    }

    fn createIdentity(
        self: Marmot,
        def_relays: []const [*c]const u8,
        boot_relays: []const [*c]const u8,
    ) MarmotError!*c.MarmotAccountSummary {
        var out: ?*c.MarmotAccountSummary = null;
        try mapStatus(c.marmot_create_identity(
            self.client,
            def_relays.ptr,
            def_relays.len,
            boot_relays.ptr,
            boot_relays.len,
            &out,
        ));
        return out.?;
    }
};

// --- reporting helpers ----------------------------------------------------

fn fail(what: []const u8) noreturn {
    std.debug.print("smoke(zig): FAILED: {s}\n", .{what});
    std.process.exit(1);
}

fn expect(condition: bool, comptime what: []const u8) void {
    if (!condition) fail(what);
    std.debug.print("smoke(zig): ok: {s}\n", .{what});
}

// --- markdown walk (DEEP: reach into the tagged-union bodies) ------------

// Reconstruct the plain text of an inline run by concatenating its Text
// spans (and wrapping Code spans in backticks) — a walk of the inline
// tagged union, mirroring print_inline_text in smoke.c.
fn printInlineText(inlines: [*c]const c.MarmotMarkdownInline, len: usize) void {
    std.debug.print("\"", .{});
    for (inlines[0..len]) |inl| {
        if (inl.tag == c.MARMOT_MARKDOWN_INLINE_TEXT) {
            std.debug.print("{s}", .{inl.unnamed_0.TEXT.content});
        } else if (inl.tag == c.MARMOT_MARKDOWN_INLINE_CODE) {
            std.debug.print("`{s}`", .{inl.unnamed_0.CODE.content});
        }
    }
    std.debug.print("\"", .{});
}

// Classify each top-level block by its tag and reach into the union body,
// mirroring walk_markdown in smoke.c. Returns the heading count seen.
fn walkMarkdown(doc: *const c.MarmotMarkdownDocument) usize {
    var headings: usize = 0;
    for (doc.blocks[0..doc.blocks_len], 0..) |block, i| {
        if (block.tag == c.MARMOT_MARKDOWN_BLOCK_HEADING) {
            headings += 1;
            const heading = block.unnamed_0.HEADING;
            std.debug.print("smoke(zig):   block {d}: heading (h{d}) ", .{ i, heading.level });
            printInlineText(heading.inlines, heading.inlines_len);
            std.debug.print("\n", .{});
        } else if (block.tag == c.MARMOT_MARKDOWN_BLOCK_PARAGRAPH) {
            std.debug.print(
                "smoke(zig):   block {d}: paragraph ({d} inlines)\n",
                .{ i, block.unnamed_0.PARAGRAPH.inlines_len },
            );
        } else if (block.tag == c.MARMOT_MARKDOWN_BLOCK_CODE_BLOCK) {
            const info = block.unnamed_0.CODE_BLOCK.info;
            if (info == null) {
                std.debug.print("smoke(zig):   block {d}: code block (lang=)\n", .{i});
            } else {
                std.debug.print("smoke(zig):   block {d}: code block (lang={s})\n", .{ i, info });
            }
        } else if (block.tag == c.MARMOT_MARKDOWN_BLOCK_LIST_BLOCK) {
            std.debug.print(
                "smoke(zig):   block {d}: list with {d} items\n",
                .{ i, block.unnamed_0.LIST_BLOCK.items_len },
            );
        } else {
            std.debug.print("smoke(zig):   block {d}: other\n", .{i});
        }
    }
    return headings;
}

pub fn main() void {
    var args = std.process.args();
    _ = args.next(); // argv[0]
    const home = args.next() orelse {
        std.debug.print("usage: smoke <fresh-home-dir>\n", .{});
        std.process.exit(2);
    };

    // --- argument validation + NULL-free discipline (no client) ---------
    const no_relays = [_][*c]const u8{};
    if (Marmot.open(null, &no_relays)) |m| {
        m.deinit();
        fail("NULL root_path should have been rejected");
    } else |err| {
        expect(err == MarmotError.NullPointer, "NULL root_path rejected");
        expect(last_error_len > 0, "NULL root_path sets last-error");
    }

    // NULL is a no-op for every free function.
    c.marmot_client_free(null);
    c.marmot_string_free(null);
    c.marmot_account_summary_list_free(null);
    c.marmot_markdown_document_free(null);
    std.debug.print("smoke(zig): ok: NULL frees are no-ops\n", .{});

    // --- construct + lifecycle ------------------------------------------
    const relays = [_][*c]const u8{"wss://relay.example.org"};
    const client = Marmot.open(home.ptr, &relays) catch {
        // Headless environments may lack a platform keychain; a documented
        // limitation, not an ABI defect.
        std.debug.print("smoke(zig): SKIP: client_new failed: {s}\n", .{lastError()});
        return;
    };
    defer client.deinit();
    std.debug.print("smoke(zig): ok: client constructed\n", .{});

    const stopping_before = client.isStopping() catch |e| fail(@errorName(e));
    expect(!stopping_before, "client not stopping");

    // start() may fail offline (dial-safety rejects unreachable relays);
    // that still exercises the start + error path. Either outcome is fine.
    if (client.start()) {
        std.debug.print("smoke(zig): ok: client started\n", .{});
    } else |_| {
        std.debug.print("smoke(zig): ok: client start offline: {s}\n", .{lastError()});
    }

    // --- Markdown: parse + DEEP walk of the recursive DTO tree ----------
    const md =
        "# Marmot\n\n" ++
        "A **bold** claim with `code` and a [link](https://example.org).\n\n" ++
        "- [x] dig burrow\n" ++
        "- [ ] store acorns\n\n" ++
        "```rust\nfn main() {}\n```\n";
    const doc = client.parseMarkdown(md) catch |e| fail(@errorName(e));
    defer c.marmot_markdown_document_free(doc);
    expect(!doc.truncated, "markdown not truncated");
    expect(doc.blocks_len >= 4, "markdown has heading + paragraph + list + code");
    const headings = walkMarkdown(doc);
    expect(headings == 1, "walked tree found exactly one heading");

    // --- offline reads --------------------------------------------------
    const accounts = client.listAccounts() catch |e| fail(@errorName(e));
    defer c.marmot_account_summary_list_free(accounts);
    expect(accounts.len == 0, "fresh home has no accounts");

    // A directory lookup for an unknown id resolves to absent (NULL out).
    const unknown_id = "0000000000000000000000000000000000000000000000000000000000000000";
    const npub = client.npub(unknown_id) catch |e| fail(@errorName(e));
    std.debug.print("smoke(zig): ok: npub lookup for absent id resolves\n", .{});
    c.marmot_string_free(npub);

    // --- error taxonomy (hard asserts) ----------------------------------
    if (client.nip65Relays("no-such-account")) |list| {
        c.marmot_string_list_free(list);
        fail("nip65_relays on unknown account should error");
    } else |err| {
        expect(err == MarmotError.UnknownAccount, "unknown account -> UnknownAccount");
    }

    if (client.revealNsec("no-such-account")) |nsec| {
        c.marmot_string_free(nsec);
        fail("reveal_nsec on unknown account should error");
    } else |err| {
        expect(err == MarmotError.UnknownAccount, "reveal_nsec unknown account -> UnknownAccount");
    }

    // --- best-effort identity (needs a relay) ---------------------------
    if (client.createIdentity(&relays, &relays)) |summary| {
        std.debug.print("smoke(zig): ok: identity created: {s}\n", .{summary.account_id_hex});
        c.marmot_account_summary_free(summary);
    } else |_| {
        std.debug.print("smoke(zig): ok: identity create offline: {s}\n", .{lastError()});
    }

    // --- shutdown -------------------------------------------------------
    client.shutdown() catch |e| fail(@errorName(e));
    std.debug.print("smoke(zig): ok: client shutdown\n", .{});
    const stopping_after = client.isStopping() catch |e| fail(@errorName(e));
    expect(stopping_after, "client reports stopping");

    std.debug.print("smoke(zig): all checks passed\n", .{});
}

File: crates/marmot-c/examples/smoke.nim
# End-to-end smoke test for the marmot-c ABI from Nim.
#
# Nim compiles to C, so it pulls in the real marmot.h and lets the C compiler
# see the true declarations (importc + header). Struct layout and signatures
# are therefore the C compiler's, not hand-transcribed. This exercises the
# same realistic slice as examples/smoke.c: client lifecycle, the last-error
# and typed-error paths, a DEEP walk of the recursive Markdown DTO tree
# (block/inline tagged unions), offline reads, a directory lookup, and the
# best-effort identity publish — all behind an idiomatic Nim wrapper that
# raises `MarmotError` and frees via `=destroy`.
#
# Build & run (see c-smoke.sh for the canonical invocation):
#   nim c -d:release --nimcache:<tmp> \
#       --passC:"-I<crate>/include -Wno-incompatible-pointer-types" \
#       --passL:"<target>/debug/libmarmot_c.a -lm -lpthread -ldl" \
#       -o:smoke_nim examples/smoke.nim
#   ./smoke_nim <fresh-empty-home-dir>
#
# `-Wno-incompatible-pointer-types`: Nim models `cstring` as `char*` and has
# no `const`-pointee notion, so it emits `char**` where the header declares
# `const char *const *relay_urls`. gcc 14+ makes that const mismatch an error
# by default; it is benign for FFI, so we downgrade just that diagnostic.

import std/[os, options]

# The header include path (-I <crate>/include) and the link flags are supplied
# on the compile command; see the build recipe above and c-smoke.sh.

# Pull declarations straight from the header via importc, so struct layout and
# signatures are the C compiler's, not hand-transcribed. cbindgen emits the
# tagged unions as `{ tag; union { HEADING; PARAGRAPH; ... }; }`; because these
# are importc + header types, Nim never computes the layout itself — it defers
# to the C header — so we only declare the union members we actually read, each
# with its C name, and access `blk.heading` compiles to `blk.HEADING`, which C
# resolves through the anonymous union.
{.push header: "marmot.h".}

type
  MarmotClient {.importc: "MarmotClient", incompleteStruct.} = object
  MarmotStringList {.importc: "MarmotStringList", incompleteStruct.} = object

  MarmotStatus {.importc: "MarmotStatus".} = distinct cint
  MarmotMarkdownBlockTag {.importc: "MarmotMarkdownBlock_Tag".} = distinct cint
  MarmotMarkdownInlineTag {.importc: "MarmotMarkdownInline_Tag".} = distinct cint

  # --- inline nodes (leaf bodies we reconstruct heading text from) -----
  MarmotMarkdownInlineTextBody {.importc: "MarmotMarkdownInline_Text_Body".} = object
    content {.importc: "content".}: cstring
  MarmotMarkdownInlineCodeBody {.importc: "MarmotMarkdownInline_Code_Body".} = object
    content {.importc: "content".}: cstring
  MarmotMarkdownInline {.importc: "MarmotMarkdownInline".} = object
    tag {.importc: "tag".}: MarmotMarkdownInlineTag
    text {.importc: "TEXT".}: MarmotMarkdownInlineTextBody
    code {.importc: "CODE".}: MarmotMarkdownInlineCodeBody

  # --- block bodies (only the fields this walk touches) ----------------
  MarmotMarkdownHeadingBody {.importc: "MarmotMarkdownBlock_Heading_Body".} = object
    level {.importc: "level".}: uint8
    inlines {.importc: "inlines".}: ptr UncheckedArray[MarmotMarkdownInline]
    inlines_len {.importc: "inlines_len".}: csize_t
  MarmotMarkdownParagraphBody {.importc: "MarmotMarkdownBlock_Paragraph_Body".} = object
    inlines {.importc: "inlines".}: ptr UncheckedArray[MarmotMarkdownInline]
    inlines_len {.importc: "inlines_len".}: csize_t
  MarmotMarkdownCodeBlockBody {.importc: "MarmotMarkdownBlock_CodeBlock_Body".} = object
    info {.importc: "info".}: cstring
    content {.importc: "content".}: cstring
  MarmotMarkdownListBlockBody {.importc: "MarmotMarkdownBlock_ListBlock_Body".} = object
    items_len {.importc: "items_len".}: csize_t
  MarmotMarkdownBlock {.importc: "MarmotMarkdownBlock".} = object
    tag {.importc: "tag".}: MarmotMarkdownBlockTag
    heading {.importc: "HEADING".}: MarmotMarkdownHeadingBody
    paragraph {.importc: "PARAGRAPH".}: MarmotMarkdownParagraphBody
    codeBlock {.importc: "CODE_BLOCK".}: MarmotMarkdownCodeBlockBody
    listBlock {.importc: "LIST_BLOCK".}: MarmotMarkdownListBlockBody

  MarmotMarkdownDocument {.importc: "MarmotMarkdownDocument".} = object
    blocks: ptr UncheckedArray[MarmotMarkdownBlock]
    blocks_len: csize_t
    truncated: bool

  MarmotAccountSummary {.importc: "MarmotAccountSummary".} = object
    label {.importc: "label".}: cstring
    account_id_hex {.importc: "account_id_hex".}: cstring
    local_signing {.importc: "local_signing".}: bool
    signed_out {.importc: "signed_out".}: bool
    running {.importc: "running".}: bool

  MarmotAccountSummaryList {.importc: "MarmotAccountSummaryList".} = object
    items: ptr MarmotAccountSummary
    len: csize_t

var
  MARMOT_STATUS_OK {.importc, nodecl.}: MarmotStatus
  MARMOT_STATUS_NULL_POINTER {.importc, nodecl.}: MarmotStatus
  MARMOT_STATUS_UNKNOWN_ACCOUNT {.importc, nodecl.}: MarmotStatus
  MARMOT_MARKDOWN_BLOCK_HEADING {.importc, nodecl.}: MarmotMarkdownBlockTag
  MARMOT_MARKDOWN_BLOCK_PARAGRAPH {.importc, nodecl.}: MarmotMarkdownBlockTag
  MARMOT_MARKDOWN_BLOCK_CODE_BLOCK {.importc, nodecl.}: MarmotMarkdownBlockTag
  MARMOT_MARKDOWN_BLOCK_LIST_BLOCK {.importc, nodecl.}: MarmotMarkdownBlockTag
  MARMOT_MARKDOWN_INLINE_TEXT {.importc, nodecl.}: MarmotMarkdownInlineTag
  MARMOT_MARKDOWN_INLINE_CODE {.importc, nodecl.}: MarmotMarkdownInlineTag

proc marmot_client_new(root_path: cstring, relay_urls: ptr cstring,
                       relay_urls_len: csize_t, out_client: ptr ptr MarmotClient): MarmotStatus {.importc.}
proc marmot_client_start(client: ptr MarmotClient): MarmotStatus {.importc.}
proc marmot_client_shutdown(client: ptr MarmotClient): MarmotStatus {.importc.}
proc marmot_client_is_stopping(client: ptr MarmotClient, out_stopping: ptr bool): MarmotStatus {.importc.}
proc marmot_client_free(client: ptr MarmotClient) {.importc.}
proc marmot_last_error_message(): cstring {.importc.}
proc marmot_string_free(s: cstring) {.importc.}
proc marmot_parse_markdown(client: ptr MarmotClient, text: cstring,
                           out_document: ptr ptr MarmotMarkdownDocument): MarmotStatus {.importc.}
proc marmot_markdown_document_free(document: ptr MarmotMarkdownDocument) {.importc.}
proc marmot_list_accounts(client: ptr MarmotClient,
                          out_list: ptr ptr MarmotAccountSummaryList): MarmotStatus {.importc.}
proc marmot_account_summary_list_free(list: ptr MarmotAccountSummaryList) {.importc.}
proc marmot_account_summary_free(summary: ptr MarmotAccountSummary) {.importc.}
proc marmot_account_nip65_relays(client: ptr MarmotClient, account_ref: cstring,
                                 out_list: ptr ptr MarmotStringList): MarmotStatus {.importc.}
proc marmot_string_list_free(list: ptr MarmotStringList) {.importc.}
proc marmot_npub(client: ptr MarmotClient, account_id_hex: cstring,
                 out_npub: ptr cstring): MarmotStatus {.importc.}
proc marmot_reveal_nsec(client: ptr MarmotClient, account_ref: cstring,
                        out_nsec: ptr cstring): MarmotStatus {.importc.}
proc marmot_create_identity(client: ptr MarmotClient,
                            default_relays: ptr cstring, default_relays_len: csize_t,
                            bootstrap_relays: ptr cstring, bootstrap_relays_len: csize_t,
                            out_summary: ptr ptr MarmotAccountSummary): MarmotStatus {.importc.}

{.pop.}

proc `==`(a, b: MarmotStatus): bool {.borrow.}
proc `==`(a, b: MarmotMarkdownBlockTag): bool {.borrow.}
proc `==`(a, b: MarmotMarkdownInlineTag): bool {.borrow.}

# --------------------------------------------------------------------------
# Idiomatic wrapper: a `Marmot` handle that raises `MarmotError` on any
# non-OK status and frees the client on destruction. The raw out-pointer
# dance stays hidden inside these procs.
# --------------------------------------------------------------------------

type
  MarmotError = object of CatchableError
    status: MarmotStatus

  Marmot = object
    client: ptr MarmotClient

  # RAII owner for a parsed document, freed as one root by `=destroy`.
  Markdown = object
    doc: ptr MarmotMarkdownDocument

proc `=copy`(dst: var Marmot, src: Marmot) {.error.}
proc `=copy`(dst: var Markdown, src: Markdown) {.error.}

proc `=destroy`(m: Marmot) =
  if m.client != nil:
    marmot_client_free(m.client)

proc `=destroy`(m: Markdown) =
  if m.doc != nil:
    marmot_markdown_document_free(m.doc)

# Take and free the thread-local detail for the most recent failure.
proc lastError(): string =
  let msg = marmot_last_error_message()
  result = if msg != nil: $msg else: "(no detail)"
  marmot_string_free(msg)

# Raise `MarmotError` (carrying the status code + last-error detail) unless
# `status` is OK. The single choke point every wrapper proc funnels through.
proc check(status: MarmotStatus, context: string) =
  if status != MARMOT_STATUS_OK:
    var e = newException(MarmotError, context & ": " & lastError())
    e.status = status
    raise e

proc open(root: string, relays: openArray[string]): Marmot =
  # `allocCStringArray` copies each URL into C-owned memory that outlives the
  # call. Borrowing `r.cstring` from the loop variable would dangle once the
  # temporary Nim string is freed under ARC/ORC.
  let arr = allocCStringArray(relays)
  defer: deallocCStringArray(arr)
  var client: ptr MarmotClient = nil
  check(marmot_client_new(root.cstring, cast[ptr cstring](arr), csize_t(relays.len), addr client),
        "client_new")
  result.client = client

proc isStopping(m: Marmot): bool =
  var stopping = true
  check(marmot_client_is_stopping(m.client, addr stopping), "is_stopping")
  stopping

# Best-effort start: OK online, a typed error offline. Returns the message on
# the offline path so the caller can report the outcome without failing.
proc tryStart(m: Marmot): Option[string] =
  let st = marmot_client_start(m.client)
  if st == MARMOT_STATUS_OK: none(string) else: some(lastError())

proc parseMarkdown(m: Marmot, text: string): Markdown =
  var doc: ptr MarmotMarkdownDocument = nil
  check(marmot_parse_markdown(m.client, text.cstring, addr doc), "parse_markdown")
  result.doc = doc

proc truncated(md: Markdown): bool = md.doc.truncated
proc blockCount(md: Markdown): int = md.doc.blocks_len.int

# Reconstruct the plain text of an inline run by concatenating its Text and
# Code spans — a small walk of the inline tagged union.
proc inlineText(inlines: ptr UncheckedArray[MarmotMarkdownInline], n: csize_t): string =
  for i in 0 ..< n.int:
    let inl = inlines[i]
    if inl.tag == MARMOT_MARKDOWN_INLINE_TEXT:
      result.add $inl.text.content
    elif inl.tag == MARMOT_MARKDOWN_INLINE_CODE:
      result.add "`" & $inl.code.content & "`"

# Walk the top-level blocks, classifying each by tag and reaching into the
# union body; reconstruct heading text from its inlines. Returns the heading
# count seen.
proc walk(md: Markdown): int =
  let doc = md.doc
  for i in 0 ..< doc.blocks_len.int:
    let blk = doc.blocks[i]
    if blk.tag == MARMOT_MARKDOWN_BLOCK_HEADING:
      inc result
      let h = blk.heading
      echo "smoke(nim):   block ", i, ": heading (h", h.level, ") \"",
           inlineText(h.inlines, h.inlines_len), "\""
    elif blk.tag == MARMOT_MARKDOWN_BLOCK_PARAGRAPH:
      let p = blk.paragraph
      echo "smoke(nim):   block ", i, ": paragraph \"",
           inlineText(p.inlines, p.inlines_len), "\""
    elif blk.tag == MARMOT_MARKDOWN_BLOCK_CODE_BLOCK:
      let c = blk.codeBlock
      echo "smoke(nim):   block ", i, ": code block (lang=",
           (if c.info != nil: $c.info else: ""), ")"
    elif blk.tag == MARMOT_MARKDOWN_BLOCK_LIST_BLOCK:
      echo "smoke(nim):   block ", i, ": list with ", blk.listBlock.items_len, " items"
    else:
      echo "smoke(nim):   block ", i, ": other"

proc accountCount(m: Marmot): int =
  var accounts: ptr MarmotAccountSummaryList = nil
  check(marmot_list_accounts(m.client, addr accounts), "list_accounts")
  result = accounts.len.int
  marmot_account_summary_list_free(accounts)

# Directory lookup: an absent id resolves to OK with a NULL string.
proc npub(m: Marmot, accountIdHex: string): Option[string] =
  var outNpub: cstring = nil
  check(marmot_npub(m.client, accountIdHex.cstring, addr outNpub), "npub")
  if outNpub == nil:
    result = none(string)
  else:
    result = some($outNpub)
    marmot_string_free(outNpub)

proc nip65Relays(m: Marmot, accountRef: string): seq[string] =
  var lst: ptr MarmotStringList = nil
  check(marmot_account_nip65_relays(m.client, accountRef.cstring, addr lst), "nip65_relays")
  # Reached only on success; the smoke run always exercises the error path.
  marmot_string_list_free(lst)

proc revealNsec(m: Marmot, accountRef: string): string =
  var outNsec: cstring = nil
  check(marmot_reveal_nsec(m.client, accountRef.cstring, addr outNsec), "reveal_nsec")
  result = if outNsec != nil: $outNsec else: ""
  marmot_string_free(outNsec)

# Best-effort identity publish: returns the account id hex online, or raises
# the typed error offline/rejected.
proc createIdentity(m: Marmot, relays: openArray[string]): string =
  # Owned C-string copies that outlive the call (see `open`).
  let arr = allocCStringArray(relays)
  defer: deallocCStringArray(arr)
  let relayPtr = cast[ptr cstring](arr)
  var summary: ptr MarmotAccountSummary = nil
  check(marmot_create_identity(m.client, relayPtr, csize_t(relays.len),
                               relayPtr, csize_t(relays.len), addr summary),
        "create_identity")
  result = if summary.account_id_hex != nil: $summary.account_id_hex else: ""
  marmot_account_summary_free(summary)

proc shutdown(m: Marmot) =
  check(marmot_client_shutdown(m.client), "shutdown")

# --------------------------------------------------------------------------

proc expect(cond: bool, what: string) =
  if not cond:
    stdout.writeLine("smoke(nim): FAILED: " & what)
    quit(1)
  stdout.writeLine("smoke(nim): ok: " & what)

proc main() =
  if paramCount() != 1:
    stderr.writeLine("usage: smoke_nim <fresh-home-dir>")
    quit(2)
  let home = paramStr(1)

  # --- argument validation + NULL-free discipline (no client) --------
  block:
    var client: ptr MarmotClient = nil
    let st = marmot_client_new(nil, nil, 0, addr client)
    expect(st == MARMOT_STATUS_NULL_POINTER, "NULL root_path rejected")

  # NULL is a no-op for every free function.
  marmot_client_free(nil)
  marmot_string_free(nil)
  marmot_account_summary_list_free(nil)
  marmot_markdown_document_free(nil)
  stdout.writeLine("smoke(nim): ok: NULL frees are no-ops")

  # --- construct + lifecycle -----------------------------------------
  var marmot: Marmot
  try:
    marmot = open(home, ["wss://relay.example.org"])
  except MarmotError as e:
    # Headless environments may lack a platform keychain; a documented
    # limitation, not an ABI defect.
    stdout.writeLine("smoke(nim): SKIP: " & e.msg)
    quit(0)
  expect(marmot.client != nil, "client constructed")

  expect(not marmot.isStopping(), "client not stopping")

  # start() may fail offline (dial-safety rejects unreachable relays); that
  # still exercises the start + error path. Either outcome is fine.
  let startErr = marmot.tryStart()
  if startErr.isNone:
    stdout.writeLine("smoke(nim): ok: client started")
  else:
    stdout.writeLine("smoke(nim): ok: client start returned offline error: " & startErr.get)

  # --- markdown: parse + DEEP walk of the recursive DTO tree ---------
  const md =
    "# Marmot\n\n" &
    "A **bold** claim with `code` and a [link](https://example.org).\n\n" &
    "- [x] dig burrow\n" &
    "- [ ] store acorns\n\n" &
    "```rust\nfn main() {}\n```\n"
  let doc = marmot.parseMarkdown(md)
  expect(not doc.truncated, "markdown not truncated")
  expect(doc.blockCount >= 4, "markdown has heading + paragraph + list + code")
  expect(doc.walk() == 1, "walked tree found exactly one heading")

  # --- offline reads -------------------------------------------------
  expect(marmot.accountCount() == 0, "fresh home has no accounts")

  # Directory lookup: the call resolves with OK (a well-formed id encodes to
  # an npub; a malformed one yields absent). Either way it must not raise.
  let unknownId = "0000000000000000000000000000000000000000000000000000000000000000"
  discard marmot.npub(unknownId)
  expect(true, "npub lookup call succeeds")

  # --- error taxonomy (hard asserts on typed status) -----------------
  try:
    discard marmot.nip65Relays("no-such-account")
    expect(false, "nip65 on unknown account should raise")
  except MarmotError as e:
    expect(e.status == MARMOT_STATUS_UNKNOWN_ACCOUNT,
           "unknown account -> MARMOT_STATUS_UNKNOWN_ACCOUNT")

  try:
    discard marmot.revealNsec("no-such-account")
    expect(false, "reveal_nsec on unknown account should raise")
  except MarmotError as e:
    expect(e.status == MARMOT_STATUS_UNKNOWN_ACCOUNT,
           "reveal_nsec on unknown account -> UNKNOWN_ACCOUNT")

  # --- best-effort identity (needs a relay) --------------------------
  try:
    let id = marmot.createIdentity(["wss://relay.example.org"])
    stdout.writeLine("smoke(nim): ok: identity created: " & id)
  except MarmotError as e:
    stdout.writeLine("smoke(nim): ok: identity create offline: " & e.msg)

  # --- shutdown ------------------------------------------------------
  marmot.shutdown()
  expect(marmot.isStopping(), "client reports stopping")

  # `marmot` (and its client) is freed by `=destroy` at scope exit.
  stdout.writeLine("smoke(nim): all checks passed")

main()
